core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -4223,6 +4243,114 @@ private void handleScalarSubQuery(
+   * count in Select list.
+   *
+   * @param parentSelect        Base SqlSelect
+   * @param selectItem          Child select items from select list
+   * @param expandedSelectItems Select items after processing
+   * @param aliases             Built from user or system values
+   * @param fields              Built up entries for each select list entry
+   */
+  private void handleTableFunctionInSelect(
+      SqlSelect parentSelect,
+      SqlBasicCall selectItem,
+      List<SqlNode> expandedSelectItems,
+      Set<String> aliases,
+      List<Map.Entry<String, RelDataType>> fields) {
+    SqlBasicCall functionCall = (SqlBasicCall) selectItem.getOperands()[0];
+    SqlFunction function = (SqlFunction) functionCall.getOperator();
+    // Check whether there are more than one table function in select list.
+    for (SqlNode item : parentSelect.getSelectList()) {
+      if (SqlUtil.isTableFunctionInSelect(item)
+          && item != selectItem) {
+        throw newValidationError(parentSelect.getSelectList(),
+            RESOURCE.onlyOneTableFunctionAllowedInSelect());
+      }
+    }
+
+    // Change the function category to USER_DEFINED_TABLE_FUNCTION.
+    // It is because that in sql-select list, the SqlFunctionCategory is USER_DEFINED_FUNCTION
+    // for a SqlUnresolvedFunction,so we should do this change.
+    if (function instanceof SqlUnresolvedFunction) {
+      if (!function.getFunctionType().isTableFunction()) {
+        SqlFunction newFunction =
+            new SqlUnresolvedFunction(function.getNameAsId(),
+                function.getReturnTypeInference(),
+                function.getOperandTypeInference(),
+                function.getOperandTypeChecker(),
+                function.getParamTypes(),
+                SqlFunctionCategory.USER_DEFINED_TABLE_FUNCTION);
+        functionCall.setOperator(newFunction);
+        function = newFunction;
+      }
+    }
+    // Check functionCall whether is a table function
+    List<SqlOperator> overloads  = new ArrayList<>();
+    opTab.lookupOperatorOverloads(function.getNameAsId(),
+        function.getFunctionType(),
+        function.getSyntax(), overloads, catalogReader.nameMatcher());
+    if (overloads.size() == 0) {
+      throw newValidationError(functionCall,
+          RESOURCE.exceptTableFunction(function.getName()));
+    }
+    // Check the parent select whether is a aggregate statement
+    if (isAggregate(parentSelect)) {
+      throw newValidationError(functionCall,
+          RESOURCE.notAllowTableFunctionInAggregate());
+    }
+    SqlNodeList aliasNodes
+        = (SqlNodeList) selectItem.getOperands()[1];
+    List<String> aliasList = new ArrayList<>();
+    for (SqlNode aliasNode : aliasNodes) {
+      aliasList.add(deriveAlias(aliasNode, aliasList.size()));
+    }
+    aliases.addAll(aliasList);
+
+    String tableAlias = TABLE_FUNCTION_PREFIX + nextGeneratedId++;
+    // Expand the table function alias
```

Comment:
Can we just use the function name as the alias prefix, the `TABLE_FUNCTION_PREFIX ` seems unnecessary.

Reply:
+1

Reply:
Thanks for your suggestion. done!